rootston/layer_shell.c

```diff
@@ -0,0 +1,181 @@
+#include <assert.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <wayland-server.h>
+#include <wlr/types/wlr_box.h>
+#include <wlr/types/wlr_surface.h>
+#include <wlr/types/wlr_layer_shell.h>
+#include <wlr/util/log.h>
+#include "rootston/desktop.h"
+#include "rootston/layers.h"
+#include "rootston/output.h"
+#include "rootston/server.h"
+
+static void apply_exclusive(struct wlr_box *output_area,
+		uint32_t anchor, uint32_t exclusive) {
+	struct {
+		uint32_t anchors;
+		int *value;
+		int multiplier;
+	} edges[] = {
+		{
+			.anchors =
+				ZWLR_LAYER_SURFACE_V1_ANCHOR_LEFT |
+				ZWLR_LAYER_SURFACE_V1_ANCHOR_RIGHT |
+				ZWLR_LAYER_SURFACE_V1_ANCHOR_TOP,
+			.value = &output_area->y,
+			.multiplier = 1,
+		},
+		{
+			.anchors =
+				ZWLR_LAYER_SURFACE_V1_ANCHOR_LEFT |
+				ZWLR_LAYER_SURFACE_V1_ANCHOR_RIGHT |
+				ZWLR_LAYER_SURFACE_V1_ANCHOR_BOTTOM,
+			.value = &output_area->height,
+			.multiplier = -1,
+		},
+		{
+			.anchors =
+				ZWLR_LAYER_SURFACE_V1_ANCHOR_LEFT |
+				ZWLR_LAYER_SURFACE_V1_ANCHOR_TOP |
+				ZWLR_LAYER_SURFACE_V1_ANCHOR_BOTTOM,
+			.value = &output_area->x,
+			.multiplier = 1,
+		},
+		{
+			.anchors =
+				ZWLR_LAYER_SURFACE_V1_ANCHOR_RIGHT |
+				ZWLR_LAYER_SURFACE_V1_ANCHOR_TOP |
+				ZWLR_LAYER_SURFACE_V1_ANCHOR_BOTTOM,
+			.value = &output_area->width,
+			.multiplier = -1,
+		},
+	};
+	for (size_t i = 0; i < sizeof(edges) / sizeof(edges[0]); ++i) {
+		if ((anchor & edges[i].anchors)) {
+			edges[i].value += exclusive * edges[i].multiplier;
+		}
+	}
+}
+
+static void arrange_layer(struct wlr_output *output, struct wl_list *list) {
+	struct wlr_box output_area = { .x = 0, .y = 0 };
+	wlr_output_effective_resolution(output,
+			&output_area.width, &output_area.height);
+	struct roots_layer_surface *roots_surface;
+	wl_list_for_each(roots_surface, list, link) {
+		struct wlr_layer_surface *layer = roots_surface->layer_surface;
+		struct wlr_layer_surface_state *state = &layer->current;
+		struct wlr_box box = { .width = state->width, .height = state->height };
+		// Horizontal axis
+		const uint32_t both_horiz = ZWLR_LAYER_SURFACE_V1_ANCHOR_LEFT
+			| ZWLR_LAYER_SURFACE_V1_ANCHOR_RIGHT;
+		if ((state->anchor & both_horiz) && box.width == -1) {
+			box.x = 0;
+			box.width = output_area.width;
+		} else if ((state->anchor & ZWLR_LAYER_SURFACE_V1_ANCHOR_LEFT)) {
+			box.x = output_area.x;
+		} else if ((state->anchor & ZWLR_LAYER_SURFACE_V1_ANCHOR_RIGHT)) {
+			box.x = output_area.width - box.width;
+		} else {
+			box.x = (output_area.width / 2) - (box.width / 2);
```

Comment:
If box.width is -1 here it's a protocol error but not sure how best to express that. Probably should have wlr_layer_shell enforce this on commit but it's annoying